examples/pytorch/mpnn/mpnn.py

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn import Parameter
+from torch.utils.data import dataloader
+import dgl
+import dgl.function as fn
+import dgl.nn.pytorch as dgl_nn
+
+import os
+from os import listdir
+from os.path import isfile, join
+import networkx as nx
+import numpy as np
+import pickle
+
+
+class NNConvLayer(nn.Module):
+    """
+    MPNN Conv Layer from Section.5 in the paper "Neural Message Passing for Quantum Chemistry"
+    """
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 edge_net,
+                 root_weight=True,
+                 bias=True):
+        """
+        Args:
+            in_channels: number of input channels
+            out_channels: number of output channels
+            edge_net: the network modules process the edge info
+            root_weight: whether add the root node feature to output
+            bias: whether add bias to the output
+        """
+
+        super().__init__()
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.edge_net = edge_net
+
+        if root_weight:
+            self.root = Parameter(torch.Tensor(in_channels, out_channels))
+        else:
+            self.register_parameter('root', None)
+
+        if bias:
+            self.bias = Parameter(torch.Tensor(out_channels))
+        else:
+            self.register_parameter('bias', None)
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        if self.root is not None:
+            nn.init.xavier_normal_(self.root.data, gain=1.414)
+        if self.bias is not None:
+            self.bias.data.zero_()
+        for m in self.edge_net.modules():
+            if isinstance(m, nn.Linear):
+                nn.init.xavier_normal_(m.weight.data, gain=1.414)
+
+    def message(self, edges):
+        return {
+            'm':
+            torch.matmul(edges.src['h'].unsqueeze(1),
+                         edges.data['w']).squeeze(1)
+        }
+
+    def apply_node_func(self, nodes):
+        aggr_out = nodes.data['aggr_out']
+        if self.root is not None:
+            aggr_out = torch.mm(nodes.data['h'], self.root) + aggr_out
+
+        if self.bias is not None:
+            aggr_out = aggr_out + self.bias
+
+        return {'h': aggr_out}
+
+    def forward(self, g, h, e):
+        h = h.unsqueeze(-1) if h.dim() == 1 else h
+        e = e.unsqueeze(-1) if e.dim() == 1 else e
+
+        g.ndata['h'] = h
+        g.edata['w'] = self.edge_net(e).view(-1, self.in_channels,
+                                             self.out_channels)
+        g.update_all(self.message, fn.sum("m", "aggr_out"),
```

Comment:
Currently DGL can not fuse user-defined message function and built-in reduce function, please try `u_mul_e` as the message function instead.

Reply:
Note that we have supported [broadcasting](https://docs.scipy.org/doc/numpy-1.15.0/user/basics.broadcasting.html) in built-in functions. For your case, first convert `g.ndata['h']` to a tensor of shape`(-1, self.in_channels, 1)` by `g.ndata['h'] = h.unsqueeze(-1)`, then call `g.update_all(fn.u_mul_e('h', 'w', 'm'), fn.sum('m', 'aggr_out'))`. This would result in a `g.ndata['aggr_out']` of shape `(-1, self.in_channels, self.out_channels)`. You may reduce the axis 1 at following apply node functions.